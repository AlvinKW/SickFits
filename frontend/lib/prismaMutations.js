import gql from 'graphql-tag';

export const ADD_TO_CART_MUTATION = gql`
	mutation ADD_TO_CART_MUTATION($id: ID!) {
		addToCart(id: $id) {
			id
			quantity
		}
	}
`;

export const CREATE_ITEM_MUTATION = gql`
	mutation CREATE_ITEM_MUTATION (
		$title: String!
		$description: String!
		$image: String
		$largeImage: String
		$price: Int!
	) {
		createItem(
			title: $title
			description: $description
			image: $image
			largeImage: $largeImage
			price: $price
		) {
			id
		}
	}
`;

export const CREATE_ORDER_MUTATION = gql`
	mutation CREATE_ORDER_MUTATION($token: String!) {
		createOrder(token: $token) {
			id
			total
			charge
			items {
				id
				title
			}
		}
	}
`;

export const DELETE_ITEM_MUTATION = gql`
	mutation DELETE_ITEM_MUTATION($id: ID!) {
		deleteItem(id: $id) {
			id
		}
	}
`;

export const REMOVE_FROM_CART_MUTATION = gql`
	mutation REMOVE_FROM_CART_MUTATION($id: ID!) {
		removeFromCart(id: $id) {
			id
		}
	}
`;

export const REQUEST_RESET_MUTATION = gql`
	mutation REQUEST_RESET_MUTATION($email: String!) {
		requestReset(email: $email) {
			message
		}
	}
`;

export const RESET_PASSWORD_MUTATION = gql`
	mutation RESET_PASSWORD_MUTATION($resetToken: String!, $password: String!, $confirmPassword: String!) {
		resetPassword(resetToken: $resetToken, password: $password, confirmPassword: $confirmPassword) {
			id
			name
			email
		}
	}
`;

export const SIGN_IN_MUTATION = gql`
	mutation SIGNIN_MUTATION($email: String!, $password: String!) {
		signIn(email: $email, password: $password) {
			id
			name
			email
		}
	}
`;

export const SIGN_OUT_MUTATION = gql`
	mutation SIGN_OUT_MUTATION {
		signOut {
			message
		}
	}
`;

export const SIGN_UP_MUTATION = gql`
	mutation SIGNUP_MUTATION($name: String!, $email: String!, $password: String!) {
		signUp(name: $name, email: $email, password: $password) {
			id
			name
			email
		}
	}
`;

export const TOGGLE_CART_MUTATION = gql`
	mutation TOGGLE_CART_MUTATION {
		toggleCart @client
	}
`;

export const UPDATE_ITEM_MUTATION = gql`
	mutation UPDATE_ITEM_MUTATION (
		$id: ID!
		$title: String
		$description: String
		$price: Int
	) {
		updateItem(
			id: $id
			title: $title
			description: $description
			price: $price
		) {
			id
			title
			description
			price
		}
	}
`;

export const UPDATE_PERMISSIONS_MUTATION = gql`
	mutation UPDATE_PERMISSIONS_MUTATION($permissions: [Permission], $userID: ID!) {
		updatePermissions(permissions: $permissions, userID: $userID) {
			id
			name
			email
			permissions
		}
	}
`;
