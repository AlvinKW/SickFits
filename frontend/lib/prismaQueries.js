import gql from 'graphql-tag';

import { perPage } from '../config';

export const ALL_ITEMS_QUERY = gql`
	query ALL_ITEMS_QUERY($skip: Int = 0, $first: Int = ${perPage}) {
		items(orderBy: createdAt_DESC, skip: $skip, first: $first) {
			id
			title
			description
			image
			largeImage
			price
		}
	}
`;

export const ALL_USERS_QUERY = gql`
	query ALL_USERS_QUERY {
		users {
			id
			name
			email
			permissions
		}
	}
`;

export const CURRENT_USER_QUERY = gql`
	query CURRENT_USER_QUERY {
		me {
			id
			name
			email
			permissions
			orders {
				id
			}
			cart {
				id
				quantity
				item {
					id
					title
					description
					image
					price
				}
			}
		}
	}
`;

export const LOCAL_STATE_QUERY = gql`
	query LOCAL_STATE_QUERY {
		cartOpen @client
	}
`;

export const PAGINATION_QUERY = gql`
	query PAGINATION_QUERY {
		itemsConnection {
			aggregate {
				count
			}
		}
	}
`;

export const SEARCH_ITEMS_QUERY = gql`
	query SEARCH_ITEMS_QUERY($searchTerm: String!) {
		items(where: { OR: [{ title_contains: $searchTerm }, { description_contains: $searchTerm }] }) {
			id
			title
			image
		}
	}
`;

export const SINGLE_ITEM_QUERY = gql`
	query SINGLE_ITEM_QUERY($id: ID!) {
		item(where: { id: $id }) {
			id
			title
			description
			largeImage
			price
		}
	}
`;

export const SINGLE_ORDER_QUERY = gql`
	query SINGLE_ORDER_QUERY($id: ID!) {
		order(id: $id) {
			id
			total
			charge
			createdAt
			items {
				id
				title
				description
				image
				price
				quantity
			}
			user {
				id
			}
		}
	}
`;

export const USER_ORDERS_QUERY = gql`
	query USER_ORDERS_QUERY {
		orders(orderBy: createdAt_DESC) {
			id
			total
			createdAt
			items {
				id
				title
				description
				image
				price
				quantity
			}
		}
	}
`;
