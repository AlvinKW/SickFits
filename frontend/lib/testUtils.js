import casual from 'casual';

casual.seed(777);

const mockItem = () => ({
	__typename: 'Item',
	id: 'abc123',
	price: 5000,
	user: null,
	image: 'dog-small.jpg',
	title: 'Dogs are the best!',
	description: 'Dogs',
	largeImage: 'dog.jpg',
});

const mockUser = () => ({
	__typename: 'User',
	id: '4234',
	name: casual.name,
	email: casual.email,
	permissions: ['ADMIN'],
	orders: [],
	cart: [],
});

const mockOrderItem = () => ({
	__typename: 'OrderItem',
	id: casual.uuid,
	image: `${casual.word}.jpg`,
	title: casual.words(),
	price: 4234,
	quantity: 1,
	description: casual.words(),
});

const mockOrder = () => ({
	__typename: 'Order',
	id: 'ord123',
	charge: 'ch_123',
	total: 40000,
	items: [mockOrderItem(), mockOrderItem()],
	createdAt: '2018-04 - 06T19: 24: 16.000Z',
	user: mockUser(),
});

const mockCartItem = overrides => ({
	__typename: 'CartItem',
	id: 'omg123',
	quantity: 3,
	item: mockItem(),
	user: mockUser(),
	...overrides,
});

class MockLocalStorage {
	constructor() {
		this.store = {};
	}

	clear() {
		this.store = {};
	}

	getItem(key) {
		return this.store[key] || null;
	}

	setItem(key, value) {
		this.store[key] = value.toString();
	}

	removeItem(key) {
		delete this.store[key];
	}
}

export {
	MockLocalStorage,
	mockItem,
	mockUser,
	mockCartItem,
	mockOrder,
	mockOrderItem,
};
