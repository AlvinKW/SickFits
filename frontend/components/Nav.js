import Link from 'next/link';
import { Mutation } from 'react-apollo';
import styled from 'styled-components';

import { TOGGLE_CART_MUTATION } from '../lib/prismaMutations';

import User from './User';
import CartCount from './CartCount';
import SignOut from './SignOut';

const StyledNav = styled.ul`
	display: flex;
	justify-self: end;
	margin: 0;
	padding: 0;
	font-size: 2rem;

	@media (max-width: 1300px) {
		justify-content: center;
		width: 100%;
		border-top: 1px solid ${props => props.theme.lightGrey};
		font-size: 1.5rem;
	}

	a, button {
		position: relative;
		display: flex;
		align-items: center;
		padding: 1rem 3rem;
		border: 0;
		font-family: "radnika_next";
		font-size: 1em;
		font-weight: 900;
		text-transform: uppercase;
		background: none;
		color: ${props => props.theme.black};
		cursor: pointer;

		@media (max-width: 700px) {
			padding: 0 10px;
			font-size: 10px;
		}

		&:before {
			content: '';
			position: absolute;
			top: 0;
			bottom: 0;
			left: 0;
			height: 100%;
			width: 2px;
			background: ${props => props.theme.lightGrey};
			transform: skew(-20deg);
		}

		&:after {
			content: '';
			position: absolute;
			left: 50%;
			height: 2px;
			width: 0;
			margin-top: 2rem;
			background: red;
			transform: translateX(-50%);
			transition: width 0.4s;
			transition-timing-function: cubic-bezier(1, -0.65, 0, 2.31);
		}

		&:hover, &:focus {
			outline: none;

			@media (max-width: 700px) {
				width: calc(100% - 10px);
			}

			&:after {
				width: calc(100% - 60px);
			}
		}
	}
`;

function Nav() {
	return (
		<User>
			{({ data: { me } }) => (
				<StyledNav>
					<Link href="/items">
						<a>Shop</a>
					</Link>
					{me && (
						<>
							<Link href="/sell">
								<a>Sell</a>
							</Link>
							<Link href="/orders">
								<a>Orders</a>
							</Link>
							<Mutation mutation={TOGGLE_CART_MUTATION}>
								{toggleCart => (
									<button onClick={toggleCart}>
										My Cart
										<CartCount
											count={me.cart.reduce((tally, cartItem) => {
												return tally + cartItem.quantity;
											}, 0)}
										/>
									</button>
								)}
							</Mutation>
							<SignOut />
						</>
					)}
					{!me && (
						<Link href="/auth">
							<a>Sign In</a>
						</Link>
					)}
				</StyledNav>
			)}
		</User>
	);
}

export default Nav;
