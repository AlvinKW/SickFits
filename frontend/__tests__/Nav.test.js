import { mount } from 'enzyme';
import wait from 'waait';
import { MockedProvider } from 'react-apollo/test-utils';
import toJSON from 'enzyme-to-json';

import { mockUser, mockCartItem } from '../lib/testUtils';
import { CURRENT_USER_QUERY } from '../lib/prismaQueries';
import Nav from '../components/Nav';

const notSignedInMocks = [
	{
		request: {
			query: CURRENT_USER_QUERY,
		},
		result: {
			data: { me: null },
		},
	},
];

const signedInMocks = [
	{
		request: {
			query: CURRENT_USER_QUERY,
		},
		result: {
			data: { me: mockUser() },
		},
	},
];

const signedInWithCartItemsMocks = [
	{
		request: {
			query: CURRENT_USER_QUERY,
		},
		result: {
			data: { me: {
				...mockUser(),
				cart: [
					mockCartItem(),
					mockCartItem(),
					mockCartItem(),
				],
			} },
		},
	},
];

describe('<Nav />', () => {
	it('renders a minimal nav when signed out', async () => {
		const wrapper = mount(
			<MockedProvider mocks={notSignedInMocks}>
				<Nav />
			</MockedProvider>
		);

		await wait();
		wrapper.update();
		const nav = wrapper.find('ul[data-test="nav"]');
		expect(toJSON(nav)).toMatchSnapshot();
	});

	it('renders a full nav when signed in', async () => {
		const wrapper = mount(
			<MockedProvider mocks={signedInMocks}>
				<Nav />
			</MockedProvider>
		);

		await wait();
		wrapper.update();
		const nav = wrapper.find('ul[data-test="nav"]');
		expect(nav.children().length).toBe(5);
		expect(nav.text()).toContain('Sign Out');
	});

	it('renders the amount of items in the cart', async () => {
		const wrapper = mount(
			<MockedProvider mocks={signedInWithCartItemsMocks}>
				<Nav />
			</MockedProvider>
		);

		await wait();
		wrapper.update();
		const nav = wrapper.find('ul[data-test="nav"]');
		const count = nav.find('div.count');
		expect(toJSON(count)).toMatchSnapshot();
	});
});
