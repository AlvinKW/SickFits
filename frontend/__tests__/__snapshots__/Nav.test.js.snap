// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Nav /> renders a minimal nav when signed out 1`] = `
<ul
  className="Nav__StyledNav-sc-1h4lqeb-0 hRvpot"
  data-test="nav"
>
  <Link
    href="/items"
  >
    <a
      href="/items"
      onClick={[Function]}
    >
      Shop
    </a>
  </Link>
  <Link
    href="/auth"
  >
    <a
      href="/auth"
      onClick={[Function]}
    >
      Sign In
    </a>
  </Link>
</ul>
`;

exports[`<Nav /> renders the amount of items in the cart 1`] = `
<div
  className="count CartCount__Dot-xxvp4g-1 hYKhWH"
>
  9
</div>
`;
